### ISSUE TITLE
Use driver's pool instead of knex one

### ISSUE THREAD

**Issue by beltschatsar** [knex/knex#2665]
# Environment

Knex version: 0.14.2
Database + version: Oracle 11gR2
OS: linux

@atiertant 

knex recycle connection in a pool so to get efficient resource usage. So far so good. Now with dialect driver's getting rich in functonnality, they offer not only pool management but other feature (proxy authentication) which mix well with it (see https://github.com/oracle/node-oracledb/blob/master/doc/api.md#pool-proxy-authentication). 

See also #2660 for the feature request about proxy authentication.

So, I would propose to delegate pool management to the driver's dialect like so (pool config following):
pool: {
  delegateToDriversDialect: true,
  min: 1
  max: 10
  options: {
   ...
  }
}

In this example, the dialect would translate poolMin/poolMax to the correct driver's property (for oracle poolMin and poolMax). options would be for dialect-specific options, or driver-specific options. 

Would you accept a pr in that way ?

Thanks

**Comment by beltschatsar** [comment]
Each dialect would need to expose theses callback in case delegateToDriversDialect is true: 
   - acquirePoolConnection(): return a promise.
   - initializePool(): return nothing. Initialize the pool.
   - destroyPool(cb): Would call the callback cb if provided, otherwise return promise
   - releaseConnectionPool(connection): Return promise
   - validateConnectionPool: Really used ?

That's all for now... I would prefer waiting for comment before starting. 

Anyone ?

**Comment by beltschatsar** [comment]
from the roadmap, it’s says:
´We’d like to use individual databases’ pooling implementations wherever possible. Each database has different error scenarios and it would be better to delegate determinations on error handling to the individual database drivers.
Our hope is to write a single package that node-mysql and node-postgres (and other database drivers) will adopt. Knex will continue to manage a pool directly via that package for clients that do not provide a pool API.´

Can we discuss about what is needed for this pool API ? @tgriesser

**Comment by beltschatsar** [comment]
I have found more updated info in [REFACTOR.md](https://github.com/tgriesser/knex/blob/refactor/REFACTOR.md#ability-to-use-knex-without-a-connection-pool--with-clustered-pools):

> Folks have asked for this before, and I think creating a convention for supplying a config option for acquiring raw driver connections / supplying custom pools will do the trick here.

so pool options like this would be ok (in a afterCreate manner) ?

```javascript
{
  pool: {
    min: 1,
    max: 4,
    insteadOfCreatePool: function (driver, driverDbConfig, done) {
      // in this example we use oracle driver's connection API
      // for that driverDbConfig is in fact oracleDbConfig and
      // driver is in fact oracledb.
      //
      // We create the pool with an application/service account 
      // (which has already been provided in user/password fields of connection config).
      // This is required because we want to impersonate the end-user
      // when doing the driver.getConnection() call (homogenous=false)
      driverDbConfig.homogeneous = false;
      driver.createPool ( driverDbConfig,
      function(err, pool) {
        if (err) {
          // failed, return error and don't try to go any further...
          // Should we do error handling here or in the caller ?
          done(err, pool);
        } else {
            // if err is not falsy, pool is available so return it
            done(err, pool);
        }
      })
    },
    insteadOfReleasePool: function (pool, done) {
      pool.close(function(Error err) {
        if (err) {
          // failed, return error and don't try to go any further...
          // Should we do error handling here or in the caller ?
          done(err, pool);
        } else {
            // if err is not falsy, pool has been released
            done(err, pool);
        }
      });
    },
    // optionsFromQueryBuilder represents params passed to .options() method of QueryBuilder
    insteadOfGetConnection: function (pool, optionsFromQueryBuilder, done) {
      pool.getConnection(optionsFromQueryBuilder, // contain impersonated user
      function (err, conn) {
        if (err) {
          // failed, return error and don't try to go any further...
          // Should we do error handling here or in the caller ?
          done(err, pool);
        } else {
            // connection is now for user HR
            // select user from dual would now return 'hr'
            done(err, pool);
        }
      });
    }
  }
}
```

would go with a usage like:
```javascript
knex
.select('user')
.from('dual')
.options({ user: 'enduser' }); // to impersonate user 'enduser'
```
This would need some modifications to support, but this should be worth it! 

I'm ready to start working on this. Let me know if you need anything else.

**Comment by beltschatsar** [comment]
In the example above, there is a insteadOfReleaseConnection missing and that would be there too.

**Comment by elhigu** [comment]
This actually sounds kind of interesting and maybe viable way to handle this issue... however there will be problems for example with some timeout implementations and how to react on those. 

I was earlier thinking that we could have just extended tarn to support delegating pool handling to underlying driver and still take advantages of timeouts implemented in tarn.

However I've been fixing so many pool problems of different drivers lately that I'm really hesitant of giving anyone else to write code related to that part currently and right now I still haven't setup / ran stress tests for oracle and I still need to add usage of transactions and transactions with timeouts to my testbench...

I would say that we should postpone this until the all current dialects are running stable and then look more if this is viable way to go.

Anyways I like the flexibility of this idea of the implementation. One thing that should be really really well defined is how those functions should behave exactly. Few months ago it wasn't consistent even internally in knex.

**Comment by beltschatsar** [comment]
Can you elaborate on those timing issues ? If possible, I would like to code test to simulate those, so to demonstrate that the new code behave as expected. 

I agree with you that the new spec should be sufficiently clear so to be implemented in other dialects.

I’m not sure if we should have tarn share his pool responsibility with the driver’s pool implementation.

For example, homogeneous pool vs heterogeneous would affect how tarn should handle resource. There is also other features that should be taking into account (read only pool vs those who support read/write). This would lead tarn to be not so obvious, and be not just a simple implementation.

And since knex using tarn, we have to set poolmin to 0 because tarn let his resource die when poolmin>0 (which was not the case with previous pool implementation). Since tarn validate each resource before borrowing them, tarn just expecting us (knex dialect) to give him the right driver’s API call to validate this, which isn’t the case right now. This expose us to connection storm which can lead to experiment database issue.

**Comment by beltschatsar** [comment]
How about keeping tarn pool then ? I have something that seems to work by keeping tarn pool while having pool delegated to dialect.

If you agree with that, I can detail how those functions should behave exactly.

**Comment by beltschatsar** [comment]
@elhigu is there anything I can do to help for this ?

Is it possible to add these hooks, but documents them as « experimental » ?

This would help in let them be used for special cases. 

Actually, i’m Struggling with getting my fork to be « npm install » compatible since the generated lib folder cannot be included in my branch (npm install must also  install the /lib folder).

**Comment by brandonros** [comment]
Why not just use Knex as a query builder and handle the connection on your own?

**Comment by joscelynjean** [comment]
As suggested by @brandonros , we started to handle the connection ourselves by providing it explicitly  to Knex (available since 0.16.1). This allow us to use the proxy user functionality from Oracle using Oracledb driver introduced since 2.3.0.

It works great for simple query using the query builder but is there any way to explicitly tell Knex to use a connection in a transaction? It doesn't seem to be supported yet?

**Comment by kibertoad** [comment]
@joscelynjean https://knexjs.org/#Builder-connection Is this what you are looking for?

**Comment by kibertoad** [comment]
If not, could you describe what kind of API you would like to have for it that you currently don't?

**Comment by joscelynjean** [comment]
Just to describe with code what I am trying to do.

To do a simple insert using the query builder, I am doing this (and it works great)!

`
const conn = await getConnectionWithProxyUser('some user');  // This return a Oracle connection
await knex({ client: 'oracledb' })
          .from('A_TABLE')
          .insert(object)
          .connection(conn);
`

If I try to run the exact same statement but in a transaction, like this : 

`
const conn = await getConnectionWithProxyUser('some user');
await knex({ client: 'oracledb' }).transaction(async trx => {

          trx
          .from('A_TABLE')
          .insert(object)
          .connection(conn);  // I think it should be provided outside the arrow function
});
`

I will get an error : "Unable to acquire a connection".
And I think it is OK since the connection should be provided at the transaction level, maybe like this : 

`
const conn = await getConnectionWithProxyUser('some user');
await knex({ client: 'oracledb' }).connection(conn).transaction(async trx => {
    ...
});
`

Maybe there is another way to provide my Oracle connection to execute my transaction?

**Comment by kibertoad** [comment]
@joscelynjean I'll add it to my ToDo list, thank you!

**Comment by joscelynjean** [comment]
Thank you @kibertoad , as soon as there is something available, we will test it on our side.

**Comment by FrancoisXavierNoe** [comment]
Hello, do you have any update on this issue ?
I would like to be able to use oracle connection pooling with adonis-lucid / knex.

**Comment by mercmobily** [comment]
I wasn't aware of this ticket when I created this one: #6292. It was a few years ago. @kibertoad wrote in 2021 "I think all known pooling problems for existing dialects (I'm looking at you, mssql) have been addressed by now". In the meantime, more issues came out, specifically #6241 (META: Connection and pooling issues). We will fix those first, and then move on to this/#6292. We will work out the overlap then.

But... I had the right idea, without even seeing this issue!

**Comment by tgriesser** [comment]
Yeah we should definitely come up with something to allow for this. This will allows for better interop when using other libraries directly and will reduce the amount of stuff we need to bake into the core library to get native functionality.

Ideally it'd be as simple as having something like a `connectionPool` option which would eliminate the need for a connection config.

```ts
Knex({
  client: 'pg',
  connectionPool: new pg.Pool(config)
})
```

There's only so many variants of the acquire/dispose among the different drivers that it should be simple for us to be able to normalize to an internal API we can consume. At least that's my hope 😄 

Any thoughts on that idea?

**Comment by mercmobily** [comment]
@tgriesser I still haven't managed to come mentally on top of the whole "connection management" part of Knex. But, I will humbly point out that there is still quite a number of issues in terms of connectivity  https://github.com/knex/knex/issues?q=state%3Aopen%20label%3A%22NG_meta_child%22   But, I can't really have a meaningful opinion on this part of Knex

**Comment by tgriesser** [comment]
@joscelynjean / @beltschatsar not sure if either of you are still around but if you have a chance, could you look at #6414 and see if it's along the lines of what you're after?
